File: compiler/ir.py
"""Workflow IR（中间表示）数据结构

IR 是 YAML 和执行图之间的中间层，解耦解析与代码生成。
MVP1 支持 SEQUENCE 模式（顺序执行）。
MVP2 扩展 CONDITIONAL 模式（条件分支）。
"""

from enum import Enum
from typing import Any

from pydantic import BaseModel


class StepType(str, Enum):
    """步骤类型枚举"""

    AGENT = "agent"
    TOOL = "tool"
    GATE = "gate"
    PARALLEL = "parallel"
    CONDITIONAL = "conditional"
    FALLBACK = "fallback"
    LOOP = "loop"
    ERROR_HANDLER = "error_handler"  # 错误处理节点
    SUB_WORKFLOW = "sub_workflow"  # 子工作流节点


class IRStep(BaseModel):
    """IR 步骤节点

    每个 step 描述工作流中的一个执行单元。
    MVP1 使用 agent/tool/gate 三种类型。
    MVP2 扩展 condition/on_true/on_false 支持条件分支。
    """

    id: str
    name: str
    type: StepType = StepType.AGENT
    agent: str | None = None
    tool: str | None = None
    input_mapping: dict[str, str] = {}
    output_key: str | None = None
    execution_policy: dict[str, Any] = {}
    checkpoint: dict[str, Any] = {}
    # MVP2: 条件分支字段
    condition: str | None = None   # 条件表达式，如 "${state.score >= 70}"
    on_true: str | None = None     # 条件为真时跳转的步骤ID
    on_false: str | None = None    # 条件为假时跳转的步骤ID
    # PARALLEL: 并行执行的子步骤列表
    parallel_steps: list["IRStep"] = []
    # FALLBACK: 主步骤和回退步骤
    primary: list["IRStep"] = []
    fallback: list["IRStep"] = []
    # LOOP: 循环执行
    loop_steps: list["IRStep"] = []
    max_iterations: int | None = None  # loop最大迭代次数
    exit_condition: str | None = None  # loop退出条件表达式
    loop_variable: str | None = None
    # ERROR_HANDLER / LOOP: 子步骤
    body: list["IRStep"] | None = None  # loop/error_handler的子步骤
    # SUB_WORKFLOW: 引用的workflow名称
    workflow_ref: str | None = None  # sub-workflow引用的workflow名称
    # ERROR_HANDLER: 错误处理步骤
    on_error: list["IRStep"] | None = None  # 错误处理步骤

    model_config = {"extra": "allow"}


class IRWorkflow(BaseModel):
    """IR 工作流

    YAML 解析后的中间表示，包含完整的步骤列表和元信息。
    SEQUENCE 模式下步骤按列表顺序依次执行。
    CONDITIONAL 模式下步骤可通过 condition/on_true/on_false 实现分支跳转。
    """

    id: str
    name: str
    description: str = ""
    version: str = "1.0"
    steps: list[IRStep] = []
    state_schema: dict[str, Any] = {}
    execution_policy: dict[str, Any] = {}
    checkpoint: dict[str, Any] = {}

    model_config = {"extra": "allow"}

File: packages/forgekin/workflow-compiler/src/ir.ts
/**
 * @flowforge/forgekin-workflow-compiler — 工作流 IR（中间表示）数据结构
 *
 * IR 是 YAML 和执行图之间的中间层，解耦解析与代码生成（对齐 Python `compiler/ir.py`）。
 * MVP1 支持 SEQUENCE 顺序执行；MVP2 支持 CONDITIONAL 条件分支；
 * MVP3 支持 PARALLEL / FALLBACK / LOOP。
 */

/** 步骤类型（对齐 Python StepType(str, Enum)） */
export type StepType =
  | 'agent'
  | 'tool'
  | 'gate'
  | 'parallel'
  | 'conditional'
  | 'fallback'
  | 'loop'
  | 'error_handler'
  | 'sub_workflow';

export const STEP_TYPES: readonly StepType[] = [
  'agent', 'tool', 'gate', 'parallel', 'conditional', 'fallback',
  'loop', 'error_handler', 'sub_workflow',
];

/** 是否为合法步骤类型 */
export function isStepType(value: unknown): value is StepType {
  return typeof value === 'string' && (STEP_TYPES as readonly string[]).includes(value);
}

/** IR 步骤节点（对齐 Python IRStep，pydantic extra=allow 语义 → 保留未知键） */
export interface IRStep {
  readonly id: string;
  readonly name: string;
  readonly type: StepType;
  readonly agent: string | undefined;
  readonly tool: string | undefined;
  readonly inputMapping: Record<string, string>;
  readonly outputKey: string | undefined;
  readonly executionPolicy: Record<string, unknown>;
  readonly checkpoint: Record<string, unknown>;
  /** MVP2 条件分支字段 */
  readonly condition: string | undefined;
  readonly onTrue: string | undefined;
  readonly onFalse: string | undefined;
  /** PARALLEL: 并行子步骤 */
  readonly parallelSteps: IRStep[];
  /** FALLBACK: 主步骤与回退步骤 */
  readonly primary: IRStep[];
  readonly fallback: IRStep[];
  /** LOOP: 循环子步骤 */
  readonly loopSteps: IRStep[];
  readonly maxIterations: number;
  readonly exitCondition: string | undefined;
  readonly loopVariable: string | undefined;
  /** ERROR_HANDLER / LOOP: 子步骤 */
  readonly body: IRStep[] | undefined;
  /** SUB_WORKFLOW: 引用的 workflow 名称 */
  readonly workflowRef: string | undefined;
  /** ERROR_HANDLER: 错误处理步骤 */
  readonly onError: IRStep[] | undefined;
  /** pydantic extra=allow：未知字段透传 */
  readonly extra: Record<string, unknown>;
}

/** IR 工作流（对齐 Python IRWorkflow） */
export interface IRWorkflow {
  readonly id: string;
  readonly name: string;
  readonly description: string;
  readonly version: string;
  readonly steps: IRStep[];
  readonly stateSchema: Record<string, unknown>;
  readonly executionPolicy: Record<string, unknown>;
  readonly checkpoint: Record<string, unknown>;
  readonly extra: Record<string, unknown>;
}

export interface MakeIRStepInit {
  readonly id?: string | undefined;
  readonly name?: string | undefined;
  readonly type?: StepType | undefined;
  readonly agent?: string | undefined;
  readonly tool?: string | undefined;
  readonly inputMapping?: Record<string, string> | undefined;
  readonly outputKey?: string | undefined;
  readonly executionPolicy?: Record<string, unknown> | undefined;
  readonly checkpoint?: Record<string, unknown> | undefined;
  readonly condition?: string | undefined;
  readonly onTrue?: string | undefined;
  readonly onFalse?: string | undefined;
  readonly parallelSteps?: IRStep[] | undefined;
  readonly primary?: IRStep[] | undefined;
  readonly fallback?: IRStep[] | undefined;
  readonly loopSteps?: IRStep[] | undefined;
  readonly maxIterations?: number | undefined;
  readonly exitCondition?: string | undefined;
  readonly loopVariable?: string | undefined;
  readonly body?: IRStep[] | undefined;
  readonly workflowRef?: string | undefined;
  readonly onError?: IRStep[] | undefined;
  readonly extra?: Record<string, unknown> | undefined;
}

export function makeIRStep(init: MakeIRStepInit & { index?: number }): IRStep {
  const index = init.index ?? 0;
  return {
    id: init.id ?? `step_${index}`,
    name: init.name ?? `Step ${index}`,
    type: init.type ?? 'agent',
    agent: init.agent,
    tool: init.tool,
    inputMapping: init.inputMapping ?? {},
    outputKey: init.outputKey,
    executionPolicy: init.executionPolicy ?? {},
    checkpoint: init.checkpoint ?? {},
    condition: init.condition,
    onTrue: init.onTrue,
    onFalse: init.onFalse,
    parallelSteps: init.parallelSteps ?? [],
    primary: init.primary ?? [],
    fallback: init.fallback ?? [],
    loopSteps: init.loopSteps ?? [],
    maxIterations: init.maxIterations ?? 1,
    exitCondition: init.exitCondition,
    loopVariable: init.loopVariable,
    body: init.body,
    workflowRef: init.workflowRef,
    onError: init.onError,
    extra: init.extra ?? {},
  };
}

export interface MakeIRWorkflowInit {
  readonly id?: string | undefined;
  readonly name?: string | undefined;
  readonly description?: string | undefined;
  readonly version?: string | undefined;
  readonly steps?: IRStep[] | undefined;
  readonly stateSchema?: Record<string, unknown> | undefined;
  readonly executionPolicy?: Record<string, unknown> | undefined;
  readonly checkpoint?: Record<string, unknown> | undefined;
  readonly extra?: Record<string, unknown> | undefined;
}

export function makeIRWorkflow(init: MakeIRWorkflowInit = {}): IRWorkflow {
  return {
    id: init.id ?? 'unnamed',
    name: init.name ?? 'Unnamed Workflow',
    description: init.description ?? '',
    version: init.version ?? '1.0',
    steps: init.steps ?? [],
    stateSchema: init.stateSchema ?? {},
    executionPolicy: init.executionPolicy ?? {},
    checkpoint: init.checkpoint ?? {},
    extra: init.extra ?? {},
  };
}
